### PR TITLE
[RFC] include: adc: add support for batch_mode

### PR DESCRIPTION
**Pull request by heinwessels**
Adds support for `batch_mode` as proposed in https://github.com/zephyrproject-rtos/zephyr/issues/55783. The change is quite minimal, but it allows the driver to have much more control, as it only needs to call `adc_context_on_sampling_done` once all sampling is done. This allows the driver to offload as much processing to the silicon as possible by doing all the samplings, setting the `ctx->sampling_index` to the final correct value, and then calling `adc_context_on_sampling_done`.

This PR also includes two commits which will be dropped later:
- A bugfix that was found which causes tests to fail: https://github.com/zephyrproject-rtos/zephyr/pull/56314
- The function implementation of `batch_mode` in the STM32 driver with an executing unit test. However, this is still WIP as the cleanup after batch mode is not yet successful, and breaks subsequent reads.

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -342,21 +346,40 @@ static int adc_stm32_dma_start(const struct device *dev,
...
+		 * which will trigger DMA TC interrupt, which we will use to
+		 * stop the sampling.
+		 */
+		dma_transfer_mode = LL_ADC_REG_DMA_TRANSFER_UNLIMITED;
+		LL_ADC_REG_SetContinuousMode(adc, LL_ADC_REG_CONV_CONTINUOUS);
+	} else {
+		/* ADC set to one-shot mode because all channels will
+		 * only read once by this DMA request
+		 */
+		dma_transfer_mode = LL_ADC_REG_DMA_TRANSFER_LIMITED;
```
Comment (tagunil) [inline drivers/adc/adc_stm32.c:363]: Actually it should be fine to use both LIMITED and UNLIMITED transfer mode here, as long as you don't enable continuous mode. Not all STM32 series support LIMITED mode, and you do stop the transfer in dma_callback anyway.

Comment (heinwessels) [reply drivers/adc/adc_stm32.c:363]: You are right, I was also thinking about that. Currently my UNLIMITED mode doesn't fully work yet so I'm still debugging what's going on. But will likely change it like you're suggesting to keep the driver simple.

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -316,14 +316,18 @@ static int adc_stm32_dma_start(const struct device *dev,
 	blk_cfg = &dma->dma_blk_cfg;
 
 	/* prepare the block */
-	blk_cfg->block_size = channel_count * sizeof(int16_t);
+	blk_cfg->block_size = data->channel_count * sizeof(int16_t);
+	if (ctx->options.batch_mode) {
```
Comment (tagunil) [inline drivers/adc/adc_stm32.c:320]: I'm afraid we need to limit the batch mode to those STM32 series that can stop continuous mode conversion without disabling the whole peripheral. That is, unless we could find a way to disable and reenable it quickly inside interrupt handler.

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -807,13 +830,29 @@ static void dma_callback(const struct device *dev, void *user_data,
 			 * within adc_context_start_sampling
 			 */
 			dma_stop(data->dma.dma_dev, data->dma.channel);
+
+			if (data->ctx.options.batch_mode) {
+				/* In batch mode the ADC must be stopped manually
+				 * because it's running in continuous mode.
+				 */
+				LL_ADC_REG_StopConversion(adc);
```
Comment (tagunil) [inline drivers/adc/adc_stm32.c:838]: Shouldn't we just extract the stopping sequence from adc_stm32_disable to a separate function and call it here?

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (tagunil) [reply drivers/adc/adc_stm32.c:363]: I think we could actually decouple transfer mode from batch mode and always use LIMITED for the "new" series (on these chips UNLIMITED is needed only for circular transfers which you don't use) and UNLIMITED for the "old" ones in single mode.

**Pull request closed without merging**